### PR TITLE
Resolve remaining predefined constants

### PR DESCRIPTION
**Pull request by herndlm**
Follow-up of #684

This resolves the remaining predefined constants that are configured via `dynamicConstantNames` and differ from what reflection is able to resolve.

There are a few constants that are defined via compilation flags where I was not a 100% sure if they can be empty or not.

And, of course, there would be many more constants from extensions that could be defined potentially and the resolving code could be refactored or grouped differently, but for now I wanted to keep it simple and adapt it based on the feedback I get.
Currently it's grouped/sorted like the following with links to docs
- group of core constants
- group of "other" core constants
- constants grouped by extensions which are sorted alphabetically
- constants inside groups are sorted as they occur in the linked docs

### CODE REVIEW

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1957,6 +1957,174 @@ private function resolveType(Expr $node): Type
...
+						new ConstantStringType('cli'),
+						new ConstantStringType('cli-server'),
+						new ConstantStringType('embed'),
+						new ConstantStringType('fpm-fcgi'),
+						new ConstantStringType('litespeed'),
+						new ConstantStringType('phpdbg'),
+						new IntersectionType([
+							new StringType(),
+							new AccessoryNonEmptyStringType(),
+						]),
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2024]: I guess this is not required, as it seems to be not required for the `PHP_OS_FAMILY` case?

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -1957,6 +1957,174 @@ private function resolveType(Expr $node): Type
...
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return IntegerRangeType::fromInterval(1, null);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MIN') {
+					return IntegerRangeType::fromInterval(null, -1);
+				}
+				if ($resolvedConstantName === 'PHP_INT_SIZE') {
+					return IntegerRangeType::fromInterval(1, null);
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2040]: hmm I think this is `4|8` only?

Comment (herndlm) [reply src/Analyser/MutatingScope.php]: right good point, this should not be anything else I guess. adapted

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2024]: Hard to say, in most cases not I guess, but I added it because https://www.php.net/manual/en/function.php-sapi-name.php explicitly states
> Although not exhaustive, the possible return values include apache, apache2handler, cgi (until PHP 5.3), cgi-fcgi, cli, cli-server, embed, fpm-fcgi, litespeed, phpdbg.

File: src/Analyser/MutatingScope.php [hunk 3]
```diff
@@ -1957,6 +1957,187 @@ private function resolveType(Expr $node): Type
...
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_EOL') {
+					return new UnionType([
+						new ConstantStringType("\n"),
+						new ConstantStringType("\r\n"),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return IntegerRangeType::fromInterval(1, null);
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2034]: per phpdocs this is either `2147483647` (32bit) or `9223372036854775807` (64bit)
https://www.php.net/manual/en/reserved.constants.php

same for `PHP_INT_MIN` with negative sign

File: src/Analyser/MutatingScope.php [hunk 4]
```diff
@@ -1969,16 +2150,25 @@ private function resolveType(Expr $node): Type
...
+				if ($resolvedConstantName === 'LIBXML_DOTTED_VERSION') {
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				// openssl, https://www.php.net/manual/en/openssl.constants.php
+				if ($resolvedConstantName === 'OPENSSL_VERSION_NUMBER') {
 					return IntegerRangeType::fromInterval(1, null);
 				}
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2173]: would it make sense to also add a `E_*` constants? would this be usefull for bitmask-logic bugs?

File: src/Analyser/MutatingScope.php [hunk 5]
```diff
@@ -1969,16 +2150,25 @@ private function resolveType(Expr $node): Type
...
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2174]: another idea which comes to mind: would adding magic-constant types for e.g. `__FILE__` etc. be helpfull?

see https://www.php.net/manual/en/language.constants.magic.php

File: src/Analyser/MutatingScope.php (hunk 3 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2034]: True, but I'm not sure if I can just use those values here. E.g. if I set `9223372036854775807` I guess phpstan won't run on 32 bit any more, right?

File: src/Analyser/MutatingScope.php (hunk 5 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2174]: I think it's a good idea, I wanted to add all the remaining useful constants anyways, I'll check later and just add them in this PR then.

Comment (staabm) [reply src/Analyser/MutatingScope.php:2174]: feel free to work in separate PRs.. I am just thinking out loud

File: src/Analyser/MutatingScope.php (hunk 4 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2173]: Looks like phpstan can resolve them already via reflection or so. But I added tests for them as well

File: src/Analyser/MutatingScope.php (hunk 5 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2174]: I tried to add them to the tests and found out that phpstan resolves them correctly already. Based on their usage, e.g. `__LINE__` would resolve to the line in the test file, `__FUNCTION__` would be empty as it's not inside a function and so on. I think adding them might make not much sense

File: src/Analyser/MutatingScope.php (hunk 3 above)
Comment (staabm) [reply src/Analyser/MutatingScope.php:2034]: I guess a unit test could tell. I don't know.

File: src/Analyser/MutatingScope.php [hunk 6]
```diff
@@ -1957,6 +1957,187 @@ private function resolveType(Expr $node): Type
...
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_MAJOR_VERSION') {
+					return IntegerRangeType::fromInterval(5, null);
+				}
+				if ($resolvedConstantName === 'PHP_MINOR_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
```
[the remark this answers is not in the document]
Comment (orklah) [reply src/Analyser/MutatingScope.php:1971]: We might as well limit to 6 the minor version.

With one year per version, even if we go to PHP 8.7 (which is doubtful) we wouldn't have to change this before 2027. 

I'd even limit the major version to 8 and change it in a few years

[the remark this answers is not in the document]
Comment (mvorisek) [reply src/Analyser/MutatingScope.php:1971]: maybe `9`...

File: src/Analyser/MutatingScope.php (hunk 3 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2034]: Apparently PHP automatically creates floats if the value is to be for int, so on 32 bit it would be a float. Which is fine as long as you don't assign it to an int property and have strict types on, see https://3v4l.org/f7lSo
I'm not sure what's best here, but I don't think we can hard-code the 64 bit int value as this could make problems when running phpstan on 32 bit.

File: src/Analyser/MutatingScope.php [hunk 7]
```diff
@@ -1957,6 +1957,187 @@ private function resolveType(Expr $node): Type
...
+				}
+				if ($resolvedConstantName === 'PHP_RELEASE_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
+				}
+				if ($resolvedConstantName === 'PHP_VERSION_ID') {
+					return IntegerRangeType::fromInterval(50300, null);
```
[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php]: Adapted to `int<50207, 89999>` for consistency with the other version constants. Let me know what you think. Unsure about the max there tbh.

File: src/Analyser/MutatingScope.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php:1985]: thx for the input @dktapps, I agree and also for the sake of keeping this simple I adapted the version constants back to have open max values. Also, this could still be adapted at a later time if really needed.

File: src/Analyser/MutatingScope.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php]: resolving for now, see https://github.com/phpstan/phpstan-src/pull/693#discussion_r740184706

File: src/Analyser/MutatingScope.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvorisek) [reply src/Analyser/MutatingScope.php:1985]: The max. values cannot be above 99, so we should go with it.

[the remark this answers is not in the document]
Comment (dktapps) [reply src/Analyser/MutatingScope.php:1985]: Actually they can, for the major versions.

[the remark this answers is not in the document]
Comment (mvorisek) [reply src/Analyser/MutatingScope.php:1985]: PHP version 99 will not be here in less than 10 years and we can always adjust if PHP versioning will be changed :)

I really think we should go with 99 now.

[the remark this answers is not in the document]
Comment (dktapps) [reply src/Analyser/MutatingScope.php:1985]: As I said before, I think this is bikeshedding. Whether it's 99 or unlimited makes no difference to anyone. I don't really care much what the end result is. But I think imposing upper limits for no reason doesn't make any sense.

https://imgs.xkcd.com/comics/workflow.png

**Comment by herndlm** [comment]
@ondrejmirtes sorry, don't want to bother you too much, but do you think the changes here make sense?
Please also note the discussion about version numbers at https://github.com/phpstan/phpstan-src/pull/693#discussion_r728121769 that derailed a bit ;)

**Comment by herndlm** [comment]
rebased and cleaned up a bit

**Comment by ondrejmirtes** [comment]
Hi, I really don’t want to read 64 comments. Anything worth my attention before I review? 😊

**Comment by herndlm** [comment]
Sorry for spaming you with the rebases :/
Most of it was about the PHP version constants (should we have upper limits or not). So a look/check there might be worth it. The rest is resolved.

File: src/Analyser/MutatingScope.php [hunk 8]
```diff
@@ -2008,6 +2008,177 @@ private function resolveType(Expr $node): Type
...
+						new ConstantStringType("\n"),
+						new ConstantStringType("\r\n"),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return IntegerRangeType::fromInterval(1, null);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MIN') {
+					return IntegerRangeType::fromInterval(null, -1);
+				}
```
[the remark this answers is not in the document]
Comment (staabm) [reply src/Analyser/MutatingScope.php:2089]: or alternatively we just fix this constant to 2 values, the actual value PHP_INT_MAX has on 32 bit php and 64 bit php

[the remark this answers is not in the document]
Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php:2089]: Yes, this should be those two values for 32bit/64bit.

[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2089]: a union type of those? I think the initial problem was that the 64 bit value is not settable on 32 bit as it would be an overflow already. or am I misunderstanding this?

[the remark this answers is not in the document]
Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php:2089]: Yeah, so maybe a single value on 32bit and union on 64bit.

[the remark this answers is not in the document]
Comment (orklah) [reply src/Analyser/MutatingScope.php:2089]: I wish there were stats on the share of users still running php on 32 bits... Couldn't find them if they exists

[the remark this answers is not in the document]
Comment (staabm) [reply src/Analyser/MutatingScope.php:2089]: Afaik 32 bit is commonly used on raspberry PI

[the remark this answers is not in the document]
Comment (orklah) [reply src/Analyser/MutatingScope.php:2089]: ... I actually want to see someone trying to run Psalm/PHPStan on a big codebase from a raspberry PI now! (sorry for the Off Topic)

[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php:2089]: Currently adding that 32/64 bit switch for the min/max constants  - there is no 32 bit CI server, right? :D should I create 32/64 bit test files for those 2 constants or just go with the 64 bit "default" case?

[the remark this answers is not in the document]
Comment (staabm) [reply src/Analyser/MutatingScope.php:2089]: Code like nextcloud/owncloud etc. is often run on raspi and therefore is coded 32bit compatible. 

I guess we need the unnion of both values here to allow projects to write special cases for 32bit. 
Its not about running phpstan on 32bit

[the remark this answers is not in the document]
Comment (herndlm) [reply src/Analyser/MutatingScope.php]: Agreed, don't get me wrong, rasing this and the union type was a good idea! I was also trying to find out if it currently runs on 32 bit. It showed that it runs just fine on 32 bit, but tests did not, which I fixed in #957. So now, it is fully compatible there and I won't/can't break it any more :)

Sooo, I adapted this PR, as suggested by Ondrej, to return the union type on 64 bit and the single, smaller value, on 32 bit, to not cause a fatal error there. To keep the tests running I simply added a few conditionals where needed for now.
Maybe that's fine, let me know what you all think.

TL/DR: adapted, check the 2 int size specific test files :)

File: src/Analyser/MutatingScope.php [hunk 9]
```diff
@@ -2008,6 +2009,182 @@ private function resolveType(Expr $node): Type
...
+				if ($resolvedConstantName === 'PHP_EOL') {
+					return new UnionType([
+						new ConstantStringType("\n"),
+						new ConstantStringType("\r\n"),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return PHP_INT_SIZE === 8
+						? new UnionType([new ConstantIntegerType(2147483647), new ConstantIntegerType(9223372036854775807)])
+						: new ConstantIntegerType(2147483647);
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:2088]: IMO, this should alwas be the union, as we cannot be sure when phpstan is running with 32 bit php, the code under test also will be and vice versa.

```suggestion
					return new UnionType([new ConstantIntegerType(2147483647), new ConstantIntegerType(9223372036854775807)]);
```

Comment (herndlm) [reply src/Analyser/MutatingScope.php:2088]: You're right, but this would then break running phpstan itself on 32bit.
Currently it runs there :/

Comment (staabm) [reply src/Analyser/MutatingScope.php:2088]: Ahh I see. Lgtm

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by herndlm** [comment]
thx, I hope this doesn't introduce annoying new errors 🤞

**Comment by ondrejmirtes** [comment]
I’ll /cc you if it does 😊
